crates/yttrium/src/sign/mod.rs

```diff
@@ -0,0 +1,1503 @@
+                            .as_secs() as i64
+                            + 60 * 60,
+                    ),
+                };
+
+                encoder.encode(serde_json::to_string(&data).unwrap().as_bytes())
+            };
+            let header = encoder.encode(
+                serde_json::to_string(&JwtHeader::default())
+                    .unwrap()
+                    .as_bytes(),
+            );
+            let message = format!("{header}.{claims}");
+            let signature = {
+                let data = key.sign(message.as_bytes());
+                encoder.encode(&data.to_bytes())
+            };
+            let auth = format!("{message}.{signature}");
+
+            let conn_opts =
+                ConnectionOptions::new(self.project_id.clone(), auth)
+                    .with_address(&self.relay_url);
+            conn_opts.as_url().unwrap().to_string()
+        };
+
+        let session_request_tx = self.session_request_tx.clone();
+        let (request_tx, mut request_rx) =
+            tokio::sync::mpsc::unbounded_channel();
+        let (irn_subscription_ack_tx, mut irn_subscription_ack_rx) =
+            tokio::sync::mpsc::unbounded_channel();
+
+        let handle_irn_subscription = {
+            let sessions = self.sessions.clone();
+            move |id: MessageId, sub_msg: Subscription| {
+                let session_sym_key = {
+                    let sessions = sessions.read().unwrap();
+                    let session = sessions.get(&sub_msg.data.topic).unwrap();
+                    session.session_sym_key
+                };
+
+                let decoded = BASE64
+                    .decode(sub_msg.data.message.as_bytes())
+                    .map_err(|e| {
+                        PairError::Internal(format!(
+                            "Failed to decode message: {e}"
+                        ))
+                    })
+                    .unwrap();
+                let envelope =
+                    envelope_type0::deserialize_envelope_type0(&decoded)
+                        .map_err(|e| PairError::Internal(e.to_string()))
+                        .unwrap();
+                let key = ChaCha20Poly1305::new(&session_sym_key.into());
+                let decrypted = key
+                    .decrypt(&Nonce::from(envelope.iv), envelope.sb.as_slice())
+                    .map_err(|e| PairError::Internal(e.to_string()))
+                    .unwrap();
+                let value =
+                    serde_json::from_slice::<serde_json::Value>(&decrypted)
+                        .map_err(|e| PairError::Internal(e.to_string()))
+                        .unwrap();
+                if let Some(method) = value.get("method") {
+                    if method.as_str() == Some("wc_sessionRequest") {
+                        // TODO implement relay-side request queue
+                        let request = serde_json::from_value::<
+                            SessionRequestJsonRpc,
```

Comment:
How do we want to parse responses?

Reply:
For every session update/request/extend etc we have the responses sent by the other peer aka dapp side